src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceGeneratedFileInfo.cs

```diff
@@ -21,16 +22,24 @@ internal sealed class MetadataAsSourceGeneratedFileInfo
 
         private readonly ParseOptions _parseOptions;
 
-        public MetadataAsSourceGeneratedFileInfo(string rootPath, Project sourceProject, INamedTypeSymbol topLevelNamedType)
+        public MetadataAsSourceGeneratedFileInfo(string rootPath, Project sourceProject, INamedTypeSymbol topLevelNamedType, bool allowDecompilation)
         {
             this.SourceProjectId = sourceProject.Id;
-            _parseOptions = sourceProject.ParseOptions;
             this.Workspace = sourceProject.Solution.Workspace;
-            this.LanguageName = sourceProject.Language;
+            this.LanguageName = allowDecompilation ? LanguageNames.CSharp : sourceProject.Language;
+            if (sourceProject.Language == LanguageName)
+            {
+                _parseOptions = sourceProject.ParseOptions;
+            }
+            else
+            {
+                _parseOptions = Workspace.Services.GetLanguageServices(LanguageName).GetRequiredService<ISyntaxTreeFactoryService>().GetDefaultParseOptionsWithLatestLanguageVersion();
+            }
+
             this.References = sourceProject.MetadataReferences.ToImmutableArray();
             this.AssemblyIdentity = topLevelNamedType.ContainingAssembly.Identity;
 
-            var extension = sourceProject.Language == LanguageNames.CSharp ? ".cs" : ".vb";
+            var extension = LanguageName == LanguageNames.CSharp ? ".cs" : ".vb";
 
             var directoryName = Guid.NewGuid().ToString("N");
             this.TemporaryFilePath = Path.Combine(rootPath, directoryName, topLevelNamedType.Name + extension);
```

Comment:
Doesn't this still need to have some differentiation for the decompiled vs. not?

Reply:
It does. Two levels of directory names are used when the keys don't match; this one will still get one unique GUID in the path where before it was getting two.